### PR TITLE
OCPBUGS-5073: Count endpoints per port for weights and enabling active health checks

### PR DESCRIPTION
**Pull request by gcs278**
Previously, you could have a service with two different ports and openshift-router would count both, regardless of what `route.spec.port.targetPort` was configured to. This change uses the route's targetPort when counting endpoints. This 
impacts weights for alternateBackends as well a bug in which a service with two ports could enable backend health checks, despite only having one server.

- `pkg/router/template/router.go`: Added the port parameter to getActiveEndpoints, calculateServiceWeights, and numberOfEndpoints.
- `pkg/router/template/router_test.go`: Updated unit testing for calculateServiceWeights to provide port.

### CODE REVIEW

**Comment by gcs278** [comment]
/hold
I need to test more and have discussion on potential edge cases this could impact.

File: pkg/router/template/router.go [hunk 1]
```diff
@@ -1068,11 +1068,15 @@ func (r *templateRouter) removeRouteInternal(route *routev1.Route) {
...
+func (r *templateRouter) numberOfEndpoints(id ServiceUnitKey, port string) int32 {
 	var eps = 0
 	svc, ok := r.findMatchingServiceUnit(id)
 	if ok && len(svc.EndpointTable) > eps {
-		eps = len(svc.EndpointTable)
+		for _, ep := range svc.EndpointTable {
+			if ep.Port == port {
+				eps += 1
+			}
+		}
```
Comment (Miciah) [inline pkg/router/template/router.go:1079]: Does this do the right thing when the route's `spec.port` is omitted (in which case the router uses all ports)?  `endpointsForAlias` checks for empty `PreferPort`, and I expect you need to do likewise here: https://github.com/openshift/router/blob/80242039851150ba43a7fdca541d3e03ddc3ba76/pkg/router/template/template_helper.go#L180-L192

Comment (gcs278) [reply pkg/router/template/router.go]: Thanks, I knew I was missing something, fixed and added more unit tests.

**Comment by gcs278** [comment]
Not related
/retest

File: pkg/router/template/router_test.go [hunk 2]
```diff
@@ -972,14 +990,104 @@ func TestCalculateServiceWeights(t *testing.T) {
 				suKey1: 1,
 			},
 		},
+		{
+			name:      "a single service with multiple endpoints, but no endpoint ports match the route port",
+			routePort: "9090",
```
Comment (frobware) [inline pkg/router/template/router_test.go:995]: I changed `routePort` to "port" but the test failed. Shouldn't the symbolic form also pass?

File: pkg/router/template/router_test.go [hunk 3]
```diff
@@ -991,20 +1099,22 @@ func TestCalculateServiceWeights(t *testing.T) {
...
+		t.Run(tc.name, func(t *testing.T) {
+			for suKey, eps := range tc.serviceUnits {
+				router.CreateServiceUnit(suKey)
+				router.AddEndpoints(suKey, eps)
+			}
+			endpointWeights := router.calculateServiceWeights(tc.serviceWeights, tc.routePort)
+			if !reflect.DeepEqual(endpointWeights, tc.expectedWeights) {
+				t.Errorf("test %s: expected endpointWeights to be %v, got %v", tc.name, tc.expectedWeights, endpointWeights)
+			}
+			// Remove endpoints and service units so the same sample template router
```
[the remark this answers is not in the document]
Comment (gcs278) [reply pkg/router/template/router_test.go:1111]: Ah yes I agree, good time to fix this. Done.

File: pkg/router/template/router_test.go (hunk 2 above)
Comment (gcs278) [reply pkg/router/template/router_test.go:995]: In this specific case, it will fail if you change it to "port" because the test matches "8080" or "port" and this test case is testing *not* matching either of those, hence why it's currently "9090". So your changing it to actually match, when we are testing for it to not match.

Comment (frobware) [reply pkg/router/template/router_test.go:993]: OK, thanks. I went through and fiddled with the inputs getting failures where I understood them and non-failures when substituting <port-number> for "port" -- except this specific case which is why I was seeking some clarification. All good.

**Comment by gcs278** [comment]
/unhold
I reviewed again, trying to look for edge cases, and can't think of any situation this would break.

File: pkg/router/template/router_test.go [hunk 4]
```diff
@@ -991,20 +1097,18 @@ func TestCalculateServiceWeights(t *testing.T) {
...
+		t.Run(tc.name, func(t *testing.T) {
+			router := NewFakeTemplateRouter()
+
+			for suKey, eps := range tc.serviceUnits {
+				router.CreateServiceUnit(suKey)
+				router.AddEndpoints(suKey, eps)
+			}
+			endpointWeights := router.calculateServiceWeights(tc.serviceWeights, tc.routePort)
+			if !reflect.DeepEqual(endpointWeights, tc.expectedWeights) {
+				t.Errorf("test %s: expected endpointWeights to be %v, got %v", tc.name, tc.expectedWeights, endpointWeights)
```
Comment (Miciah) [inline pkg/router/template/router_test.go:1109]: You don't need to include the test name in the error message; using `t.Run` gets you that for free.

Comment (gcs278) [reply pkg/router/template/router_test.go]: Done

**Comment by Miciah** [comment]
Thanks!
/approve
